### PR TITLE
Merge the FIRRTL HierPathOp with HW GlobalRefOp

### PR DESCRIPTION
**Pull request by sequencer**
The functionality of `GlobalRefOp` and `HierPathOp` share a same logic to represent a NLA path.
However `GlobalRefOp` doesn't provide enough utilities to developer.
This PR decouple `HierPathOp` to FIRRTL Dialect by adding a new interface: `HierarchicalPathLike`, making those two ops share a same Interface, thus providing more utilities to `GlobalRefOp`

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 1]
```diff
@@ -3566,7 +3566,7 @@ bool HierPathOp::hasModule(StringAttr modName) {
 }
 
 /// Return true if the NLA has the InnerSym .
-bool HierPathOp::hasInnerSym(StringAttr modName, StringAttr symName) const {
+bool HierPathOp::hasInnerSym(StringAttr modName, StringAttr symName) {
```
Comment (sequencer) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:3569]: Seems there is no good why to add `const` in ODS.

**Comment by prithayan** [comment]
Overall this seems fine, I was wondering if we can merge the two and replace `firrtl::HierPathOp` with the `hw::GlobalRef` .

**Comment by sequencer** [comment]
Basically yes, as a new contributor, I dare not removing some codes ;p
but if some maintainer think it's ok to replace `firrtl::HierPathOp` with `hw::GlobalRef`, I'll rewrite this PR.

**Comment by prithayan** [comment]
Actually merging the two ops is a task I was planning to do this week. Basically use `GlobalRef` in FIRRTL dialect also. 
If you are willing, you can create a new PR for that, else I might do it later this week. 
We can then have followup discussion on that PR.

**Comment by sequencer** [comment]
> Basically use GlobalRef in FIRRTL dialect also.

I'll create a new PR to do so.

**Comment by prithayan** [comment]
Thanks for working on this. I just glanced at the changes and they seem to look good. 
I was not sure how the verifier can be ported to the `HW` dialect, but seems like you have already done that.
Let me know when this is ready to review, of if you need any help in this.

**Comment by sequencer** [comment]
>  I was not sure how the verifier can be ported to the HW dialect, but seems like you have already done that.

Not actually, I have been thinking about this for two days...
Since `GlobalRef` is an operator for HW Dialect, so it assumes the module to be `HWModuleOp` and `HWModuleExternOp` inside HW Dialect, since for the dependence issue, it cannot cast to FIRRTL Module, so basically we should provide more utilities to `HWModuleLike` interface, and `GlobalRefOp` should only depends on the method which operate on the `HWModuleLike`.
Things might be tricky here, I thought here are the steps we need to do:
1. There are some small semantic difference between `GlobalRefOp` and `HierPathOp`, we should clarify them and merge them, for example, how to refer to the element in the top module.
2. Since `verifySymbolUses` should only operate on the `HWModuleLike` methods rather than `HWModule`, methods should be added to `HWModuleLike`.

I will firstly append some utilities to `HWModuleLike` and continue on this PR. But I personally think we can merge the `HierarchicalPathLike` firstly to ease the rebase burden.

**Comment by seldridge** [comment]
This PR is going to get beefy if we aren't careful. We can probably factor it in the following way:

1. One patch that adds all the necessary updates to HWModuleLike, etc.
2. One patch that duplciates functionality of `HierPathOp` into `GlobalRefOp`.
3. One patch that does the switchover.

**Review by darthscsi — CHANGES_REQUESTED** [review]
One thing that needs to be figured out is how to deal with the richer set of hierarchical path specifiers in SystemVerilog which don't exist in firrtl.  Namely `$unit` and `$root`.  Currently `$root` is supported in sv.xmr, but arguably it should be part of a global path op.  This is from 3.12.1 of the sv spec.  A SV-compatible hierarchical path op needs these options to be able to represent all of SV.  FIRRTL representation does not as it has an explicit top-level.

So while I have encouraged @prithayan to persue the unification path, it's not clear to me anymore that they should be unified.  They are sufficiently different, and verification and types are sufficiently different, that it is likely just easier to have an op for each world.  There is some duplicate logic, it is true, but there might be better ways to deal with that than unifying the ops or adding more op interfaces.

**Comment by darthscsi** [comment]
> it's not clear to me anymore that they should be unified

We have instead, been working on cleaning up the FIRRTL versions as much as possible.  This is why InnerSymbolTable and such traits haven't been used outside of FIRRTL: not because we don't think they should be, but because we want to understand how they should work and how verification can be done efficiently before we try to generalize.

Which is to say, resist the urge to prematurely unify across dialects just because things look similar until we have a solid understanding of how each thing is suppose to behave and what it really represents.

**Pull request closed without merging**